lib/host/wasi/inode-win.cpp

```diff
@@ -82,53 +434,310 @@ INode INode::stdErr() noexcept {
+  // this
+  // https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-openfilebyid?redirectedfrom=MSDN
+  // and this
+  // https://docs.microsoft.com/en-us/windows/win32/api/winbase/ns-winbase-file_id_info
+  FileStat.ino = 0;
+  // TODO: Find an equivalent for device ID in windows
+  FileStat.dev = 0;
+  FileStat.nlink = (*FileInfo).nNumberOfLinks;
+  FileStat.size = static_cast<uint64_t>((*FileInfo).nFileSizeLow) +
+                  (static_cast<uint64_t>((*FileInfo).nFileSizeHigh) << 32);
+  FileStat.atim =
+      (static_cast<uint64_t>((*FileInfo).ftLastAccessTime.dwHighDateTime)
+       << 32) +
+      (static_cast<uint64_t>((*FileInfo).ftLastAccessTime.dwLowDateTime)) -
+      TICKS_TO_UNIX_EPOCH;
+  FileStat.mtim =
+      (static_cast<uint64_t>((*FileInfo).ftLastWriteTime.dwHighDateTime)
+       << 32) +
+      (static_cast<uint64_t>((*FileInfo).ftLastWriteTime.dwLowDateTime)) -
+      TICKS_TO_UNIX_EPOCH;
+  FileStat.ctim =
+      (static_cast<uint64_t>((*FileInfo).ftCreationTime.dwHighDateTime) << 32) +
+      (static_cast<uint64_t>((*FileInfo).ftCreationTime.dwLowDateTime)) -
+      TICKS_TO_UNIX_EPOCH;
+
+  return {};
 }
 
-WasiExpect<void> INode::fdFilestatSetSize(__wasi_filesize_t) const noexcept {
-  return WasiUnexpect(__WASI_ERRNO_NOSYS);
+WasiExpect<void>
+INode::fdFilestatSetSize(__wasi_filesize_t Size) const noexcept {
+
+  // FIXME: This function sets the size correctly but does not fill the extra
+  // bytes with zeros
+
+  FILE_STANDARD_INFO StandardInfo;
+  FILE_ALLOCATION_INFO AllocationInfo;
+
+  if (unlikely(GetFileInformationByHandleEx(Handle, FileStandardInfo,
+                                            &StandardInfo,
+                                            sizeof(StandardInfo))) == FALSE) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  }
+
+  // Update the size attribute
+  AllocationInfo.AllocationSize = toLargeIntegerFromUnsigned(Size);
+
+  if (SetFileInformationByHandle(Handle, FileAllocationInfo, &AllocationInfo,
+                                 sizeof(AllocationInfo)) == FALSE) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  }
+
+  return {};
 }
 
-WasiExpect<void> INode::fdFilestatSetTimes(__wasi_timestamp_t,
-                                           __wasi_timestamp_t,
-                                           __wasi_fstflags_t) const noexcept {
-  return WasiUnexpect(__WASI_ERRNO_NOSYS);
+WasiExpect<void>
+INode::fdFilestatSetTimes(__wasi_timestamp_t ATim, __wasi_timestamp_t MTim,
+                          __wasi_fstflags_t FstFlags) const noexcept {
+
+  // Let FileTime be initialized to zero if the times need not be changed
+  FILETIME AFileTime = {0, 0};
+  FILETIME MFileTime = {0, 0};
+
+  // For setting access time
+  if (FstFlags & __WASI_FSTFLAGS_ATIM) {
+    uint64_t Aticks = ATim / NANOSECONDS_PER_TICK + TICKS_TO_UNIX_EPOCH;
+    AFileTime.dwLowDateTime = static_cast<DWORD>(Aticks % 0x100000000ULL);
```

Comment:
`0x100000000ULL` is in too many place, use a const instead.

Reply:
Will be using one of the time constants.